core/ir/synthetic/decode.c

```diff
@@ -0,0 +1,139 @@
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
+ * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH
+ * DAMAGE.
+ */
+
+/* decode.c -- a decoder for DR synthetic IR */
+
+#include "decode.h"
+
+#include "../globals.h"
+#include "encoding_common.h"
+#include "instr_api.h"
+#include "opnd_api.h"
+
+/* Decodes the raw bytes of an encoded instruction \p encoded_instr into DR instruction
+ * representation \p instr.
+ * Returns next instruction's pc.
+ *
+ * The smallest encoded instruction (i.e., with no operands) has 4 bytes and follows
+ * this scheme:
+ * |----------------------| |--| |----| |----|
+ * 31..               ..10  9,8   7..4   3..0
+ *       category          eflags #src  #dst
+ *
+ * 22 bits, category: holds values of type #dr_instr_category_t. Note that an instruction
+ * can belong to more than one category.
+ * 2 bits, eflags: most significant bit set to 1 indicates the instruction reads at least
+ * one arithmetic flag, least significant bit set to 1 indicates the instruction writes
+ * at least one arithmetic flag.
+ * 4 bits, #src: number of source operands (read) that are registers.
+ * 4 bits, #dst: number of destination operands (written) that are registers.
+ * (Note that we are only interested in register dependencies, hence operands that are
+ * not register, such as immediates or memory references, are not present)
+ *
+ * Instructions with operands can reach up to 14 bytes: the 4 bytes described above, and
+ * up to 10 register operands of 1 byte each.
+ * For example, an instruction with 4 operands (2 src, 2 dst) has 4 additional bytes that
+ * are encoded following this scheme:
+ * |--------| |--------| |--------| |--------|
+ * 31.. ..24  23.. ..16  15..  ..8  7..   ..0
+ * src_opnd0  src_opnd1  dst_opnd0  dst_opnd1
+ *
+ * We assume all encoded values to be little-endian.
+ */
+byte *
+decode_from_synth(dcontext_t *dcontext, byte *encoded_instr, instr_t *instr)
+{
+    uint encoding = 0;
+    uint shift = 0;
+
+    /* Copy the first 4 bytes of encoded_instr (which is always present) in a uint for
+     * easier retrieving of category, eflags, #src, #dst values.
+     */
+    memcpy(&encoding, encoded_instr, sizeof(encoding));
+
+    /* Decode number of destination operands.
+     */
+    uint num_dsts_mask = ((1U << NUM_OPND_BITS) - 1);
+    uint num_dsts = encoding & num_dsts_mask;
+    shift += NUM_OPND_BITS;
+
+    /* Decode number of source operands.
+     */
+    uint num_srcs_mask = ((1U << NUM_OPND_BITS) - 1) << shift;
+    uint num_srcs = (encoding & num_srcs_mask) >> shift;
+    shift += NUM_OPND_BITS;
+
+    instr_set_num_opnds(dcontext, instr, num_dsts, num_srcs);
+
+    /* Decode flags.
+     */
+    // TODO i#6662: need to set flags in instr_t.
```

Comment:
What is the correct way to set arithmetic flags in an instr_t?
I couldn't quite figure it out.

Reply:
I think the decoder routines directly write `instr->eflags`.

Reply:
Understood. Done.